### PR TITLE
Implement support for custom images.

### PR DESCRIPTION
**Pull request by boumenot**
This PR implements support for custom VHDs, and closes a long-standing [issue](https://github.com/Azure/packer-azure/issues/209) from the packer-azure repo.

Refactor the way the ARM template is built, by creating a builder type to handle document manipulation.

Simplify the way templates and template parameters are created.

Implemented simple [approval tests](http://approvaltests.sourceforge.net/) in Go to validate JSON document manipulations.  The reason this code change includes *.approved.txt* files is due to approval testing.  These *approved* files represent expected test outcomes, and are validated when test instances are re-executed.  It's much easier to validate against a golden copy than reparsing a document to validate specific parts, and still ensure you did not miss anything.

CC: @tasquith - I don't know how far you made it on your changes for supporting custom images.  This work pre-empts it.  It turned out to be considerably more involved than I expected.  I would appreciate it if you could try this code out, and give it a once over too.

CC: @paulmey - would you please review

### CODE REVIEW

**Comment by cbednarski** [comment]
> These *approved files represent expected test outcomes, and are validated when test instances are re-executed. It's much easier to validate against a golden copy than reparsing a document to validate specific parts, and still ensure you did not miss anything.

Great strategy!

**Comment by tomasquith** [comment]
@boumenot , more than happy to give this a once over, it looks like a much cleaner way! - I'll take a look when the CI build is clear. As far as my work, I was confident with it on linux - just windows needed a few sanity checks.

File: builder/azure/arm/builder.go [hunk 1]
```diff
@@ -83,13 +85,20 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
...
+		return GetVirtualMachineDeployment(config)
+	}
+	keyVaultDeploymentFactory := func(config *Config) (*resources.Deployment, error) {
+		return GetKeyVaultDeployment(config)
+	}
+
 	var steps []multistep.Step
 
-	if b.config.OSType == constants.Target_Linux {
+	if strings.EqualFold(b.config.OSType, constants.Target_Linux) {
```
Comment (cbednarski) [inline builder/azure/arm/builder.go]: Nice! I didn't know about EqualFold.

Comment (boumenot) [inline builder/azure/arm/builder.go]: I just learned about it too.  I came across while I was researching something else.

I also just learned about [x/net/context](https://godoc.org/golang.org/x/net/context) and [x/net/trace](https://godoc.org/golang.org/x/net/trace).  Great stuff !  I have some more refactoring to do based on this, especially around my interruptible task work.

**Comment by boumenot** [comment]
@tomasquith - there is little difference between the Windows and Linux code for this change, but the WIndows deployments are **slow**.  (I know you mentioned this too.)  I need to find someone internally to dig into this more, and figure out why.  The images are 4x the size, but the time is well over 4x.

File: builder/azure/common/approvals/approvals.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+	return state, nil
+}
+
+// Walk the call stack, and try to find the test method that was executed.
+// The test method is identified by looking for the test runner, which is
+// *assumed* to be common across all callers.  The test runner has a Name() of
+// 'testing.tRunner'.  The method immediately previous to this is the test
+// method.
+func findTestMethod() (*testState, error) {
```
Comment (cbednarski) [inline builder/azure/common/approvals/approvals.go]: If this and other functions in this file are only used in tests the file should have a `_test.go` suffix (mainly to indicate as such to other developers).

**Comment by cbednarski** [comment]
Please remember to include updated docs to explain how to use the new features (and if there are any BC breaks please call them out so I can include them in the changelog).

**Comment by tasquith** [comment]
I've just ran a few of my work scenarios for our CentOS images through, looks good to me! I'll have a look at kicking off the windows ones now!

File: builder/azure/common/approvals/approvals.go (hunk 2 above)
Comment (boumenot) [inline builder/azure/common/approvals/approvals.go]: I tried this, and it breaks the tests.  It looks like Go is ignoring _test.go files when considering what files to bring in.

```text
C:/go/v1.6.2\bin\go.exe test -v ./arm/...
go build github.com/mitchellh/packer/builder/azure/common/approvals: no buildable Go source files in C:\dev\packer-azure.go\src\github.com\mitchellh\packer\builder\azure\common\approvals
```

Comment (paulmey) [inline builder/azure/common/approvals/approvals.go]: There probably has to be at lease one none _test.go file in that directory?

File: builder/azure/arm/builder.go [hunk 3]
```diff
@@ -83,13 +85,20 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
 
 	b.setTemplateParameters(b.stateBag)
 
+	machineTemplateFactory := func(config *Config) (*resources.Deployment, error) {
+		return GetVirtualMachineDeployment(config)
+	}
```
Comment (paulmey) [inline builder/azure/arm/builder.go]: `machineTemplateFactory := GetVirtualMachineDeployment` ?

File: builder/azure/arm/builder.go [hunk 4]
```diff
@@ -83,13 +85,20 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
...
+	keyVaultDeploymentFactory := func(config *Config) (*resources.Deployment, error) {
+		return GetKeyVaultDeployment(config)
+	}
```
Comment (paulmey) [inline builder/azure/arm/builder.go]: `keyVaultDeploymentFactory := GetKeyVaultDeployment` ?

**Comment by paulmey** [comment]
LGTM

**Comment by tasquith** [comment]
@boumenot - Roughly how long were windows images from a vhd taking to provision for you?

I've got a Win2k12R2 from the marketplace, added chocolatey and installed a few packages, sysprepped, which captured fine and pretty quickly.

Using that VHD as a source for another machine (just with a powershell command to list the installed packages then re sysprep) it seems to take ages just deploying the template, although no errors are shown in the azure ui.

Did your internal team manage to shed any light on the speed?

Regards,
Tom

**Comment by boumenot** [comment]
Thank you for the ping!  I do not have more details, I'll re-ping them.

It was taking me almost 40 minutes to deploy Windows images.  It is slow comparatively speaking.

**Comment by tasquith** [comment]
Thanks! I've just managed to get it working. It was hanging because I wasn't deleting the unattend.xml file before sysprep. 

It's now taking just under 20 mins to build the images, although I'm not really doing anything taxing with the provisioners. Doesn't seem too bad!

Thanks for the help @boumenot

**Comment by boumenot** [comment]
w00t!  Thanks for following up.  I'll add this to the docs to help with troubleshooting!

**Comment by riezebosch** [comment]
It's not documented anywhere as far as I could see, but found out you should use the 'image_url' for a `vhd`: https://github.com/hashicorp/packer/blob/520dea23fd846ef72acb2d16a85cbe92f86e83a1/builder/azure/arm/config.go#L592

The other options for a managed image, managed disk or shared image gallery are also mentioned in that location.

Probably the tool would've told me if I dropped my current arguments but I took the hard way wandering through this repository.

**Comment by boumenot** [comment]
There is documentation about custom images, see https://packer.io./docs/builders/azure.html#vhd-or-managed-image.

There is an example in the repository that may be helpful too, see https://github.com/hashicorp/packer/blob/master/examples/azure/windows_custom_image.json or https://github.com/hashicorp/packer/blob/master/examples/azure/linux_custom_image.json for Linux.

**Pull request merged**